[InstCombine] fold unsigned predicates on srem result

jacobly0 opened the pull request:
This allows optimization of more signed floor implementations when the divisor is a known power of two to an arithmetic shift.

Proof for the implemented optimizations:
https://alive2.llvm.org/ce/z/j6C-Nz

Proof for the test cases:
https://alive2.llvm.org/ce/z/M_PBjw

dtcxzyw reviewed (changes requested):
I think this optimization is too specific.
For your motivating case, we can simply convert `icmp ugt (srem i32 %x, 8), -2147483648` into `icmp slt (srem i32 %x, 8), 0`. Then InstCombine will convert the remain expression into `ashr X, 3`: https://godbolt.org/z/Y7K5hd7c3

Generalized proof: https://alive2.llvm.org/ce/z/gST2A5

jacobly0 commented:
> I think this optimization is too specific. For your motivating case, we can simply convert `icmp ugt (srem i32 %x, 8), -2147483648` into `icmp slt (srem i32 %x, 8), 0`. Then InstCombine will convert the remain expression into `ashr X, 3`: https://godbolt.org/z/Y7K5hd7c3
> 
> Generalized proof: https://alive2.llvm.org/ce/z/gST2A5

Thanks for the hint, it should now be generalized to any comparison constant outside of the range of the srem result range.

dtcxzyw commented on llvm/lib/Transforms/InstCombine/InstCombineCompares.cpp:
```diff
@@ -2674,10 +2674,41 @@ Instruction *InstCombinerImpl::foldICmpShrConstant(ICmpInst &Cmp,
+    // (X % DivisorC) ugt C -> (X % DivisorC) slt 0
+    //   iff (C slt 0 ? ~C : C) uge abs(DivisorC)-1
+    // (X % DivisorC) ult C+1 -> (X % DivisorC) sgt -1
+    //   iff (C+1 slt 0 ? ~C : C) uge abs(DivisorC)->
```
```suggestion
    //   iff (C+1 slt 0 ? ~C : C) uge abs(DivisorC)
```

dtcxzyw commented on llvm/lib/Transforms/InstCombine/InstCombineCompares.cpp:
```diff
@@ -2674,10 +2674,41 @@ Instruction *InstCombinerImpl::foldICmpShrConstant(ICmpInst &Cmp,
+      NormalizedC.flipAllBits();
+    assert(!DivisorC->isZero() &&
+           "srem X, 0 should have been simplified already.");
+    if (!NormalizedC.uge(DivisorC->abs() - 1))
```
```suggestion
    if (!NormalizedC.ugt(DivisorC->abs()))
```

Proof: https://alive2.llvm.org/ce/z/mmxS8u

jacobly0 replied:
Unfortunately, that's the wrong direction (the equivalent expression is `NormalizedC.ugt(DivisorC->abs() - 2)`), so applying that diff causes the fold to no longer happen in cases where it is still valid.  As much as I tried to make the math look nicer that was already the best I could manage while still having the correct bounds.
```diff
diff --git a/llvm/lib/Transforms/InstCombine/InstCombineCompares.cpp b/llvm/lib/Transforms/InstCombine/InstCombineCompares.cpp
index c107a12fcf43..8d99b5f1022a 100644
--- a/llvm/lib/Transforms/InstCombine/InstCombineCompares.cpp
+++ b/llvm/lib/Transforms/InstCombine/InstCombineCompares.cpp
@@ -2696,7 +2696,7 @@ Instruction *InstCombinerImpl::foldICmpSRemConstant(ICmpInst &Cmp,
       NormalizedC.flipAllBits();
     assert(!DivisorC->isZero() &&
            "srem X, 0 should have been simplified already.");
-    if (!NormalizedC.uge(DivisorC->abs() - 1))
+    if (!NormalizedC.ugt(DivisorC->abs()))
       return nullptr;
 
     Type *Ty = SRem->getType();
diff --git a/llvm/test/Transforms/InstCombine/icmp-srem.ll b/llvm/test/Transforms/InstCombine/icmp-srem.ll
index 47429dbad74c..3c9a65e57fd0 100644
--- a/llvm/test/Transforms/InstCombine/icmp-srem.ll
+++ b/llvm/test/Transforms/InstCombine/icmp-srem.ll
@@ -17,7 +17,7 @@ define i1 @icmp_ugt_srem5_m5(i32 %x) {
 ; CHECK-LABEL: define i1 @icmp_ugt_srem5_m5(
 ; CHECK-SAME: i32 [[X:%.*]]) {
 ; CHECK-NEXT:    [[R:%.*]] = srem i32 [[X]], 5
-; CHECK-NEXT:    [[C:%.*]] = icmp slt i32 [[R]], 0
+; CHECK-NEXT:    [[C:%.*]] = icmp ugt i32 [[R]], -5
 ; CHECK-NEXT:    ret i1 [[C]]
 ;
   %r = srem i32 %x, 5
@@ -53,7 +53,7 @@ define i1 @icmp_ugt_srem5_4(i32 %x) {
 ; CHECK-LABEL: define i1 @icmp_ugt_srem5_4(
 ; CHECK-SAME: i32 [[X:%.*]]) {
 ; CHECK-NEXT:    [[R:%.*]] = srem i32 [[X]], 5
-; CHECK-NEXT:    [[C:%.*]] = icmp slt i32 [[R]], 0
+; CHECK-NEXT:    [[C:%.*]] = icmp ugt i32 [[R]], 4
 ; CHECK-NEXT:    ret i1 [[C]]
 ;
   %r = srem i32 %x, 5
@@ -89,7 +89,7 @@ define i1 @icmp_ult_srem5_m4(i32 %x) {
 ; CHECK-LABEL: define i1 @icmp_ult_srem5_m4(
 ; CHECK-SAME: i32 [[X:%.*]]) {
 ; CHECK-NEXT:    [[R:%.*]] = srem i32 [[X]], 5
-; CHECK-NEXT:    [[C:%.*]] = icmp sgt i32 [[R]], -1
+; CHECK-NEXT:    [[C:%.*]] = icmp ult i32 [[R]], -4
 ; CHECK-NEXT:    ret i1 [[C]]
 ;
   %r = srem i32 %x, 5
@@ -125,7 +125,7 @@ define i1 @icmp_ult_srem5_5(i32 %x) {
 ; CHECK-LABEL: define i1 @icmp_ult_srem5_5(
 ; CHECK-SAME: i32 [[X:%.*]]) {
 ; CHECK-NEXT:    [[R:%.*]] = srem i32 [[X]], 5
-; CHECK-NEXT:    [[C:%.*]] = icmp sgt i32 [[R]], -1
+; CHECK-NEXT:    [[C:%.*]] = icmp ult i32 [[R]], 5
 ; CHECK-NEXT:    ret i1 [[C]]
 ;
   %r = srem i32 %x, 5

```
https://alive2.llvm.org/ce/z/naNgve

dtcxzyw replied:
Oops. Ignore me.

dtcxzyw reviewed (commented):
Please add some negative tests.

dtcxzyw commented on llvm/test/Transforms/InstCombine/icmp-srem.ll:
Can you add some tests with negative divisors?

jacobly0 replied:
Yeah I just thought of that and was add it.  Note that the only negative constant divisor that survives other folding is smin.

dtcxzyw reviewed (approved):
LGTM. Thank you!
Please wait for additional approval from other reviewers :)

jacobly0 committed:
[InstCombine][NFC] precommit tests for signed floor division

goldsteinn commented on llvm/lib/Transforms/InstCombine/InstCombineCompares.cpp:
```diff
@@ -2674,10 +2674,41 @@ Instruction *InstCombinerImpl::foldICmpShrConstant(ICmpInst &Cmp,
+    // (X % DivisorC) ugt C -> (X % DivisorC) slt 0
+    //   iff (C slt 0 ? ~C : C) uge abs(DivisorC)-1
+    // (X % DivisorC) ult C+1 -> (X % DivisorC) sgt -1
+    //   iff (C+1 slt 0 ? ~C : C) uge abs(DivisorC)-1
```
nit: For the comment can you be consistent w/ either ~C syntax or ~IR syntax i.e


`(X % DivisorC) ugt C -> (X % DivisorC) slt 0)` -> `(icmp ugt (srem X, DivisorC), C) -> (icmp slt (srem X, DivisorC), 0)` or `(X s% DivisorC) u> C` -> `(X s% DivisorC) s< 0`

jacobly0 replied:
I was trying to copy the style of the nearby code, but I certainly wanted to write `s%` and `s<` so I'll do that.

jacobly0 committed:
[InstCombine] fold unsigned predicates to signed on srem result

This allows optimization of more signed floor implementations when the
divisor is a known power of two to an arithmetic shift.

Proof for the implemented optimizations:
https://alive2.llvm.org/ce/z/j6C-Nz

Proof for the test cases:
https://alive2.llvm.org/ce/z/M_PBjw

goldsteinn commented:
LGTM
